File: 2025-07-July-GFG-POTD/July 07 - Next Greater Element in Circular Array/Solution.py
from typing import List

class Solution:
    def nextLargerElement(self, arr: List[int]) -> List[int]:
        n = len(arr)
        res = [-1] * n
        stack = []

        for i in range(2 * n - 1, -1, -1):
            index = i % n
            while stack and arr[stack[-1]] <= arr[index]:
                stack.pop()
            if i < n:
                if stack:
                    res[index] = arr[stack[-1]]
            stack.append(index)

        return res

File: 2025-07-July-GFG-POTD/July 07 - Next Greater Element in Circular Array/Solution.java
class Solution {
   public ArrayList<Integer> nextLargerElement(int[] arr) {
        int n = arr.length;
        ArrayList<Integer> res = new ArrayList<>(Collections.nCopies(n, -1));
        Stack<Integer> stack = new Stack<>();

        // Traverse the array twice for circular behavior
        for (int i = 2 * n - 1; i >= 0; i--) {
            int index = i % n;
            // Maintain decreasing stack
            while (!stack.isEmpty() && arr[stack.peek()] <= arr[index]) {
                stack.pop();
            }
            // Only fill result in the first pass
            if (i < n) {
                if (!stack.isEmpty()) {
                    res.set(index, arr[stack.peek()]);
                }
            }
            stack.push(index);
        }
        return res;
    }
}
